source/agora/consensus/protocol/Nominator.d

```diff
@@ -944,7 +945,7 @@ extern(D):
             }
             const block = makeNewBlock(last_block,
                 externalized_tx_set, data.time_offset, random_seed,
-                this.enroll_man.getCountOfValidators(last_block.header.height + 1),
+                validators.length - data.missing_validators,
```

Comment:
I think this might have been the issue. `makeNewBlock` expects the count of validators to be the number of validators that can sign the block, but `validators.length` is the number of validators *active* (regardless of pre-image revelation) at this height.

Reply:
I think the problem is that a `validator` can be slashed at the `height` of this block when it did not reveal a `preimage`. It then does not get included in the `getValidators` so the count is less than it should be. I think we should slash at next height.